File: Problems/0303_Range_Sum_Query/Range_Sum_Query.cs
using System;
using System.Collections.Generic;
using System.Linq;



public class Solution
{
    /*
    public class NumArray {

        int[] nums;
        public NumArray(int[] nums) {
            this.nums = nums;
        }
        
        public int SumRange(int i, int j) {
            int sum = 0;
            for (int n = i; n <= j && n < nums.Length; ++n)
                sum += nums[n];
            return sum;
        }
    }
    */
    public class NumArray
    {
        private int[] sum;

        public NumArray(int[] nums)
        {
            sum = new int[nums.Length + 1];
            for (int i = 0; i < nums.Length; i++) {
                sum[i + 1] = sum[i] + nums[i];
            }
        }

        public int SumRange(int i, int j)
        {
            return sum[j + 1] - sum[i];
        }    
    }

    private int[] set_array_int(string[] flds)
    {
        if (flds.Length == 0)
            return null;
        
        int[] nums = new int[flds.Length];
        for (int i = 0; i < flds.Length; ++i)
            nums[i] = int.Parse(flds[i]);

        return nums;
    }

    private string output_array_int(int[] nums)
    {
        if (nums.Length <= 0)
            return "";

        string resultStr = nums[0].ToString();
        for (int i = 1; i < nums.Length; ++i)
            resultStr += "," + nums[i].ToString();

        return resultStr;
    }

    public void Main(string args)
    {
        string[] flds = args.Replace("[","").Replace("]","").Split(',');
        int[] nums = set_array_int(flds);
        Console.WriteLine("nums = " + output_array_int(nums));

        System.Diagnostics.Stopwatch sw = new System.Diagnostics.Stopwatch();
        sw.Start();

        NumArray obj = new NumArray(nums);
        Console.WriteLine("sumRange(0, 2) = " + obj.SumRange(0, 2));
        Console.WriteLine("sumRange(2, 5) = " + obj.SumRange(2, 5));
        Console.WriteLine("sumRange(0, 5) = " + obj.SumRange(0, 5));
        
        sw.Stop();
        Console.WriteLine("Execute time ... " + sw.ElapsedMilliseconds.ToString() + "ms");
    }
}

File: Problems/0300_0399/0303_Range_Sum_Query/Project_Python3/Range_Sum_Query.py
# coding: utf-8

import os
import sys
import time
import copy

class NumArray:
    def __init__(self, nums):
        """
        :type nums: List[int]
        """
        self.sum = [0]*(len(nums) + 1)
        for i in range(len(nums)):
            self.sum[i + 1] = self.sum[i] + nums[i]

    def sumRange(self, i, j):
        """
        :type i: int
        :type j: int
        :rtype: int
        """
        return self.sum[j + 1] - self.sum[i]

def array_str_to_int(flds):
    if len(flds) <= 0:
        return None
    nums = [0]*len(flds)
    nums[0] = int(flds[0])
    for i in range(1, len(flds)):
        nums[i] = int(flds[i])
    return nums

def main():
    argv = sys.argv
    argc = len(argv)

    if argc < 2:
        print("Usage: python {0} <testdata.txt>".format(argv[0]))
        exit(0)

    if not os.path.exists(argv[1]):
        print("{0} not found...".format(argv[1]))
        exit(0)

    testDataFile = open(argv[1], "r")
    lines = testDataFile.readlines()

    for temp in lines:
        temp = temp.strip()
        if temp == "":
            continue
        print("args = {0}".format(temp))
        loop_main(temp)
    #   print("Hit Return to continue...")
    #   input()

def loop_main(temp):
    tempStr = temp.replace("[","").replace("]","").rstrip()

    flds = tempStr.split(',')
    nums = array_str_to_int(flds)
    print("nums = {0}".format(nums))

    time0 = time.time()

    obj = NumArray(nums)
    print("sumRange(0, 2) = {0:d}".format(obj.sumRange(0, 2)))
    print("sumRange(2, 5) = {0:d}".format(obj.sumRange(2, 5)))
    print("sumRange(0, 5) = {0:d}".format(obj.sumRange(0, 5)))

    time1 = time.time()

    print("Execute time ... : {0:f}[s]\n".format(time1 - time0))

if __name__ == "__main__":
    main()
